src/serde_arrow_variable_list_array.erl

```diff
@@ -0,0 +1,83 @@
+-module(serde_arrow_variable_list_array).
+
+-behaviour(serde_arrow_array).
+
+-export([new/2]).
+
+-include("serde_arrow_array.hrl").
+-spec new(Values :: list(), Type :: map() | serde_arrow_type:arrow_type()) -> Array :: #array{}.
+new(Values, Opts) when is_map(Opts) ->
+    case maps:get(type, Opts, undefined) of
+        undefined ->
+            erlang:error(badarg);
+        Type when is_tuple(Type) orelse is_atom(Type) ->
+            new(Values, Type)
+    end;
+new(Values, GivenType) when
+    (is_tuple(GivenType) andalso tuple_size(GivenType) =:= 2) orelse is_atom(GivenType)
+->
+    Len = length(Values),
+    {Bitmap, NullCount} = serde_arrow_bitmap:validity_bitmap(Values),
+    Type = serde_arrow_type:normalize(GivenType),
+    %% Data
+    Flattened = serde_arrow_utils:flatten(Values),
+    Array = serde_arrow_fixed_primitive_array:new(Flattened, Type),
+    %% Offsets
+    [0 | FlatOffsets] = serde_arrow_offsets:new_list(Flattened, Type),
+    Offsets = serde_arrow_buffer:new([0 | offsets(Values, FlatOffsets, 0)], {s, 32}),
+    #array{
+        layout = variable_list,
+        type = Type,
+        len = Len,
+        element_len = undefined,
+        null_count = NullCount,
+        validity_bitmap = Bitmap,
+        offsets = Offsets,
+        data = Array
+    };
+new(Values, {Layout, NestedType, Size} = Type) when
+    (Layout =:= variable_list andalso Size =:= undefined) orelse
+        (Layout =:= fixed_list andalso is_integer(Size))
+->
+    Len = length(Values),
+    {Bitmap, NullCount} = serde_arrow_bitmap:validity_bitmap(Values),
+    Flattened = serde_arrow_utils:flatten(Values),
+    Array = serde_arrow_array:new(Layout, Flattened, NestedType),
+    Offsets = serde_arrow_buffer:new(nested_offsets(Values, Type), {s, 32}),
+    #array{
+        layout = variable_list,
+        type = Type,
+        len = Len,
+        element_len = undefined,
+        null_count = NullCount,
+        validity_bitmap = Bitmap,
+        offsets = Offsets,
+        data = Array
+    }.
+
+%%%%%%%%%%%
+%% Utils %%
+%%%%%%%%%%%
+
+offsets([H | T], Offsets, CurOffset) when (H =:= undefined) orelse (H =:= nil) ->
+    [CurOffset | offsets(T, Offsets, CurOffset)];
+offsets([H | T], Offsets, _CurOffset) ->
+    Len = length(H),
+    {HOffsets, TOffsets} = lists:split(Len, Offsets),
+    CurOffset = lists:last(HOffsets),
```

Comment:
`lists:last() is a be expensive. Should we split on `Len - 1` instead? Then you can access CurOffset without doing the extra work.